patient_data_source.py

```diff
@@ -0,0 +1,214 @@
+		"""Returns the patient's date of birth."""
+		return datetime.min
+
+	def get_identifier_system(self) -> str:
+		"""Returns the namespace for unique identifier values."""
+		return None
+
+	def generate_name(self, gender : Gender) -> tuple[str, str]:
+		"""Generate a first and last name based on gender."""
+		match gender:
+			case Patient.Gender.male:
+				first_name = names.get_first_name('male')
+			case Patient.Gender.female:
+				first_name = names.get_first_name('female')
+			case default:
+				first_name = names.get_first_name()
+
+		return names.get_last_name(), first_name
+
+	def get_name(self) -> tuple[str, str]:
+		"""Returns a a tuple of the last and first name of the patient. Default implementation generates names based on gender"""
+		return self.generate_name(self.get_gender())
+
+class Observation(ABC):
+	class ObservationType(Enum):
+		fio2 = 0
+		pip = 1
+		peep = 2
+		hr = 3
+		sao2 = 4
+
+	# Units used for ObservationType's valus. Codes that follow the spec in https://ucum.org/ucum.html.
+	UNIT_CODES = {
+	  ObservationType.fio2 : '',
+	  ObservationType.pip : 'cm[H20]',
+	  ObservationType.peep : 'cm[H20]',
+	  ObservationType.hr : '/min',
+	  ObservationType.sao2 : '%',
+	}
+
+	# LOINC codes that I found by using loinc.org/search/ ... and making my best guesses when things were unclear
+	# Not to be fully trusted
+	LOINC_CODES = {
+	  ObservationType.fio2 : '19996-8',
+	  ObservationType.pip : '60951-1',
+	  ObservationType.peep : '20077-4',
+	  ObservationType.hr : '8867-4',
+	  ObservationType.sao2 : '59408-5',
+	}
+
+	# Human readable strings for the ObservationTypes currently implemented.
+	DISPLAY_STRINGS = {
+	  ObservationType.fio2 : 'ETT Sx Quality',
+	  ObservationType.pip : 'PIP',
+	  ObservationType.peep : 'PEEP',
+	  ObservationType.hr : 'Heart Rate',
+	  ObservationType.sao2 : 'Sa02',
+	}
+
+	def get_identifier_value(self) -> str:
+		"""Returns a id value for external use."""
+		return None
+
+	@abstractmethod
+	def get_observation_type(self) -> ObservationType:
+		"""Returns the observation's ObservationType. Used internally for returning other Observation attribues"""
+		pass
+
+	def get_identifier_system(self) -> str:
+		"""Returns the namespace for identifier values."""
+		return None
+	
+	def get_unit_string(self) -> str:	
+		"""Returns humnan readable units for the Observation's value. Default implementation uses UNIT_CODES"""
+		return self.get_unit_code()
+
+	def get_display_string(self) -> str:
+		"""Returns a human readable description of the ObservationType."""
+		return self.DISPLAY_STRINGS[self.get_observation_type()]
+
+	def get_unit_code(self) -> str:
+		"""Returns a computer processable form for the Observation's value. Default implementation uses the UCUM system."""
+		return self.UNIT_CODES[self.get_observation_type()]
+
+	def get_loinc_code(self) -> str:
+		"""Returns a computer processable form for the ObservationType. Default implementation uses the LOINC_CODES."""
```

Comment:
Notice that we have hardcoded `http://loinc.org` as the observation coding system below. So we are not offering a choice to use a different coding system.
Therefore this docstring should say "Returns LOINC code for the ObservationType".

Alternatively, we can stop hardcoding `http://loinc.org` abd add a `Observation.get_observation_system` function that has loinc as the default implementaion

Reply:
Since the change is small, going for flexibility seems like the best way.